### ISSUE TITLE
Allowing multiple scope annotations on components

### ISSUE THREAD

**Issue by pyricau** [google/dagger#282]
Right now, AFAIK Dagger won't let me do this (`ScopeA` and `ScopeB` are two annotations annotated with `@Scope`:

```
@Component @ScopeA @ScopeB
public interface MyComponent {}
```

I get this:

```
Error: java.lang.IllegalArgumentException: MyComponent was annotated with more than one @Scope annotation
```

I don't see any good reason why this wouldn't work, please let me know if I forgot something. Now, why would anyone want to do that?

The answer is code reuse. We have an app with many scopes, and we want many sibling / cousin scopes to share some singleton binding configuration, while not sharing singleton instances.

So say I defined some shared code:

```
public interface Shared {
  Presenter presenter();
}

@Scope public @interface SharedScope {}

@SharedScope public class Presenter {
  @Inject Presenter() {}
}
```

Then I can use it on two different components:

```
@Component @SharedScope
public interface ComponentA extends Shared {
}
```

```
@Component @SharedScope
public interface ComponentB extends Shared {
}
```

But now, I can't share yet another piece of code:

```
@Component @Shared1Scope @Shared2Scope
public interface ComponentA extends Shared1, Shared2 {
}
```

We have some common views that need singleton presenters and are included in various places in the app, so that's how we've been dealing with those so far.

**Comment by pyricau** [comment]
For anyone running into this problem, it can be temporarily avoided by introducing extra empty component layers:

```java
@Component
public interface RootComponent {
  ComponentAPartial componentAPartial();
}

@Subcomponent @Shared1Scope 
public interface ComponentAPartial {
  ComponentA componentAFull();
}

@Subcomponent @Shared2Scope
public interface ComponentA extends Shared1, Shared2 {
}
```

That way you can do:

```java
ComponentA component = rootComponent.componentAPartial().componentAFull();
```

**Comment by pyricau** [comment]
Note that we should still prevent two components with an ancestor relationship to share the same scope.

**Comment by tadfisher** [comment]
This was shipped with 2.1? https://github.com/google/dagger/blob/master/CHANGELOG.md

**Comment by gk5885** [comment]
Indeed it was.
